### PR TITLE
Numexpr 2.4.6

### LINKED ISSUE #15213: DEPS: drop numexpr < 2.4.6
since 2.4.4 is specifically exluded because of some multi-threading issues, this eliminates the pandas checks for this (as well as some compat code for earlier versions of numexpr).

### PR DESCRIPTION
**Pull request by FrancescAlted**
- [x] closes #15213 
 - [x] passes ``git diff upstream/master | flake8 --diff``
 - [x] whatsnew entry

This addresses #15213.  Actually, the recommended version for me is 2.6.2, but I understand that this is too recent still.

### CODE REVIEW

**Comment by jorisvandenbossche** [comment]
@FrancescAlted we have a few occurences of numpexpr pinned at older versions in ci/requirement files (I think the 2.7_COMPAT.run and 3.4_SLOW.run ones). Best to change those versions to 2.4.6 then.

File: doc/source/install.rst [hunk 1]
```diff
@@ -226,7 +226,7 @@ Recommended Dependencies
 
 * `numexpr <https://github.com/pydata/numexpr>`__: for accelerating certain numerical operations.
   ``numexpr`` uses multiple cores as well as smart chunking and caching to achieve large speedups.
-  If installed, must be Version 2.1 or higher (excluding a buggy 2.4.4). Version 2.4.6 or higher is highly recommended.
+  If installed, must be Version 2.4.6 or higher.  Numexpr 2.6.2 or higher is strongly recommended.
```
Comment (jreback) [inline doc/source/install.rst]: you don't need the recommended part

File: doc/source/whatsnew/v0.20.0.txt [hunk 2]
```diff
@@ -1,12 +1,16 @@
...
+v0.20.0 (????, 2017)
 --------------------
 
 This is a major release from 0.19 and includes a small number of API changes, several new features,
 enhancements, and performance improvements along with a large number of bug fixes. We recommend that all
 users upgrade to this version.
 
+.. warning::
+
+   Due to important fixes and performance improvements, ``numexpr`` version is now required to be >= 2.4.6 and it will not be used at all if this requisite is not fullfilled (:issue:`15213`).
```
Comment (jreback) [inline doc/source/whatsnew/v0.20.0.txt]: i would just move this to the other api changes section; just say 2.4.6 is now required

File: pandas/computation/__init__.py [hunk 3]
```diff
@@ -7,22 +7,13 @@
...
     import numexpr as ne
     ver = ne.__version__
-    _NUMEXPR_INSTALLED = ver >= LooseVersion('2.1')
+    _NUMEXPR_INSTALLED = ver >= LooseVersion('2.4.6')
 
-    # we specifically disallow 2.4.4 as
-    # has some hard-to-diagnose bugs
-    if ver == LooseVersion('2.4.4'):
-        _NUMEXPR_INSTALLED = False
-        warnings.warn(
```
Comment (jreback) [inline pandas/computation/__init__.py]: there is a test for this i think

**Review by jreback — CHANGES_REQUESTED** [review]

File: doc/source/install.rst (hunk 1 above)
Comment (FrancescAlted) [inline doc/source/install.rst]: Well, IMO the improvements in https://github.com/pydata/numexpr/blob/master/RELEASE_NOTES.rst#changes-from-261-to-262 are worth a recommendation, but I can remove it if you insist.

File: doc/source/whatsnew/v0.20.0.txt (hunk 2 above)
Comment (FrancescAlted) [inline doc/source/whatsnew/v0.20.0.txt]: You probably mean "Other enhancements" section (numexpr does not introduce any API change).

Comment (jreback) [inline doc/source/whatsnew/v0.20.0.txt]: it's an API change for pandas meaning if u has a version that is no longer supported then things will change - its also something to be aware of when upgrading

File: doc/source/install.rst (hunk 1 above)
Comment (jreback) [inline doc/source/install.rst]: it may be but back compat is more important

File: pandas/computation/__init__.py (hunk 3 above)
Comment (FrancescAlted) [inline pandas/computation/__init__.py]: Yep.  Addressed in https://github.com/pandas-dev/pandas/pull/15383/commits/f22559876b130891e7c834cf269d1289307e3fa3.

Comment (FrancescAlted) [inline pandas/computation/__init__.py]: and fixed a typo in https://github.com/pandas-dev/pandas/pull/15383/commits/7a275ce974fbd85cbbf65f2feda32b06ea836917

**Comment by FrancescAlted** [comment]
There are a couple of ["allowed failures" in TravisCI](https://travis-ci.org/pandas-dev/pandas/builds/201215599).  Any hint on what's going on there?

**Comment by jorisvandenbossche** [comment]
For the MacOSX one, I see 

```
UnsatisfiableError: The following specifications were found to be in conflict:
  - numexpr 2.4.6
  - pytables 3.0.0 -> numexpr 2.2*
```

**Comment by FrancescAlted** [comment]
Uh, I see the Mac OSX build that completed just fine.  Also, the AppVeyor failure seems like a transitory failure during conda packages access:

```
conda.exceptions.CondaRuntimeError: Runtime error: RuntimeError: Runtime error: HTTPError: 502 Server Error: Bad Gateway for url: https://repo.continuum.io/pkgs/free/win-64/xlsxwriter-0.9.6-py35_0.tar.bz2: https://repo.continuum.io/pkgs/free/win-64/xlsxwriter-0.9.6-py35_0.tar.bz2
```

But I am still unsure on the level of importance of the ["allowed failures" in TravisCI](https://travis-ci.org/pandas-dev/pandas/builds/201215599).

**Comment by jorisvandenbossche** [comment]
Sorry, I mislooked, not the MacOSX build of course, build number 6 with the old versions (2.7_COMPAT), the error now is:

```
UnsatisfiableError: The following specifications were found to be in conflict:
  - numexpr 2.4.6 -> numpy 1.10*
  - numpy 1.7.1
Use "conda info <package>" to see the dependencies for each package.
```

We want to keep the numpy 1.7 there for now (although there is an issue to bump that requirement as well), so I would just keep the older numexpr in this build (then it is a test that you get the correct error for that).

> But I am still unsure on the level of importance of the "allowed failures" in TravisCI.

They are not 'allowed' to fail in principle (only for specific things, eg compat issue with numpy master that is not yet fixed). I think it is like that to get quicker feedback from travis.

There is still a lint error:

```
pandas/computation/__init__.py:18:17: E128 continuation line under-indented for visual indent
```

The appveyor issue indeed does not seem to be related to this PR.

**Comment by jorisvandenbossche** [comment]
@FrancescAlted Was it needed to also revert the version change in the 3.4_SLOW build? As this one was not failing I think.

**Comment by FrancescAlted** [comment]
Yes, it was: https://travis-ci.org/pandas-dev/pandas/builds/201238220.  I am not familiar with the convention to run different configurations, but I'd say that stems from: https://github.com/pandas-dev/pandas/blob/master/.travis.yml#L131.  I just reverted the numexpr versions existing in this two scenarios.  My intention is to look at the possible errors after the revert and try to fix them.

**Comment by jreback** [comment]
i would change this version to 2.4.6, otherwise lgtm. https://github.com/pandas-dev/pandas/blob/master/ci/requirements-3.4_SLOW.run

ping on green.

**Review by jreback — APPROVED** [review]

**Comment by FrancescAlted** [comment]
I think we will still have a remaining error [here](https://github.com/pandas-dev/pandas/blob/master/pandas/tests/computation/test_compat.py#L29).  But the assert will not raise the error as such, and we would need something like `from pandas.computation import __NUMEXPR_INSTALLED`.  However, there is a line at the beginning saying `from pandas.computation import _MIN_NUMEXPR_VERSION` which already raises the warning, and hence the error is not re-raised.  Not sure how to cope with that.  Opinions?

**Comment by jorisvandenbossche** [comment]
So probably that test never worked .. What you could do is move the import inside the test and in a "catch warnings" context, and then check that a warning is raised in the specific case and not otherwise (https://docs.python.org/3.5/library/warnings.html#testing-warnings)?

**Comment by jreback** [comment]
no it worked but the import process is changed by this PR
it needs to be protected

**Comment by FrancescAlted** [comment]
I have been looking at the original code, and I don't think it worked because `assert not _NUMEXPR_INSTALLED` cannot possibly raise a Warning; can it?

**Comment by jreback** [comment]
so I think with the changed up order of tests, this ``test_compat`` that is failing (on 2.7_COMPAT) is not able to assert the warning as its already been produced (you can see in the output actually).

So I guess just change the test then.

**Comment by FrancescAlted** [comment]
Ok, I think that I have found the issue and fixed it in e1b34a9.  Basically, one needs to force the `reload()` of `pandas.computation` module to trigger the `UserWarning` in import time.  Hopefully we should see all green after this one.

**Comment by FrancescAlted** [comment]
Well, it turns out that `reload()` is tricky and it still cannot raise the `UserWarning` in Python 2.7 (although 3.4 seems to work well).  After googling a bit, I have read that [`reload()` is very tricky to get right](http://stackoverflow.com/a/684229/3095325), so my vote is to remove this `UserWarning` test completely (unless you have a better proposition).

**Comment by jreback** [comment]
@FrancescAlted I would just remove the check for the warning (leave the test just don't assert that it produces)

File: pandas/tests/computation/test_compat.py [hunk 4]
```diff
@@ -10,6 +10,12 @@
 
 from pandas.computation.engines import _engines
 import pandas.computation.expr as expr
+from pandas.computation import _MIN_NUMEXPR_VERSION
+# Get reload for Python 3.4 and on, if not, use internal reload()
```
Comment (jreback) [inline pandas/tests/computation/test_compat.py]: too complicated, don't reload

File: pandas/tests/computation/test_compat.py [hunk 5]
```diff
@@ -21,12 +27,11 @@ def test_compat():
     try:
         import numexpr as ne
         ver = ne.__version__
-        if ver == LooseVersion('2.4.4'):
-            assert not _NUMEXPR_INSTALLED
-        elif ver < LooseVersion('2.1'):
+        if ver < LooseVersion(_MIN_NUMEXPR_VERSION):
             with tm.assert_produces_warning(UserWarning,
```
Comment (jreback) [inline pandas/tests/computation/test_compat.py]: just remove this ``assert_produces_warning``

File: pandas/tests/computation/test_compat.py [hunk 6]
```diff
@@ -21,12 +27,11 @@ def test_compat():
...
                                             check_stacklevel=False):
-                assert not _NUMEXPR_INSTALLED
+                reload(pd.computation)
```
Comment (jreback) [inline pandas/tests/computation/test_compat.py]: too complicated

**Comment by jreback** [comment]
lgtm. ping on green.

**Comment by FrancescAlted** [comment]
Done.  Hopefully this will give us all green (bar some temporary failures that apparently are usual in TravisCI).

**Comment by jreback** [comment]
@FrancescAlted well there have been some flakiness with some gbq tests; I turned them off for now. We *should* have all green. 

FYI we just have the fail fast to make the results appear quicker. We have LOTS of configurations to test. So allowed failures are really NOT allowed :>

**Comment by jreback** [comment]
one more FYI. every push on a PR generates a new travis run (which can take hours). So generally like to bunch them up if possible. This PR of course is a case that its necessary to test little changes so no big deal.

**Comment by jreback** [comment]
thanks @FrancescAlted

**Pull request closed without merging**